lib/components/shorts/shorts_videos_stream_player.dart

```diff
@@ -0,0 +1,154 @@
+    await _initializeControllerAtIndex(0);
+    await _initializeControllerAtIndex(1);
+    _playControllerAtIndex(0);
+    return true;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      children: [
+        PageView.builder(
+            scrollDirection: Axis.vertical,
+            itemCount: widget.shortsVideos.length,
+            onPageChanged: (pageIndex) async {
+              if (pageIndex > prevPageIndex) {
+                await _playNext(pageIndex);
+              } else {
+                await _playPrevious(pageIndex);
+              }
+              setState(() => prevPageIndex = pageIndex);
+            },
+            itemBuilder: (context, index) {
+              return FutureBuilder(
+                future: _future,
+                builder: (BuildContext context, AsyncSnapshot snapshot) {
+                  if (snapshot.connectionState == ConnectionState.waiting) {
+                    return const LoadingGeneric();
+                  } else {
+                    return (_controllers[index] != null)
+                        ? ShortsVideoPlayerScreen(
+                            shortsVideo: widget.shortsVideos[index],
+                            playControllerAtIndex: _playControllerAtIndex,
+                            stopControllerAtIndex: _stopControllerAtIndex,
+                            index: index,
+                            controller: _controllers[index]!,
+                          )
+                        : const LoadingGeneric();
+                  }
+                },
+              );
+            }),
+      ],
+    );
+  }
+
+  Future _playPrevious(int pageIndex) async {
+    _stopControllerAtIndex(pageIndex + 1);
+    _disposeControllerAtIndex(pageIndex + 2);
+    _playControllerAtIndex(pageIndex);
+    await _initializeControllerAtIndex(pageIndex - 1);
+  }
+
+  Future _playNext(int pageIndex) async {
+    _stopControllerAtIndex(pageIndex - 1);
+    _disposeControllerAtIndex(pageIndex - 2);
+    _playControllerAtIndex(pageIndex);
+    await _initializeControllerAtIndex(pageIndex + 1);
+  }
+
+  Future _initializeControllerAtIndex(int index) async {
+    if (widget.shortsVideos.length > index && index >= 0) {
+      // Create new controller
+      final VideoPlayerController controller = VideoPlayerController.network(
+        widget.shortsVideos[index].hlsUrl,
+        formatHint: VideoFormat.hls,
+        videoPlayerOptions: VideoPlayerOptions(allowBackgroundPlayback: false),
+      );
+      controller.seekTo(widget.shortsVideos[index].duration);
+      // Add to [controllers] list
+      await controller.initialize();
+      setState(() {
+        _controllers[index] = controller;
+      });
+      return controller.value.isInitialized;
+    }
+    return true;
+  }
+
+  Future _playControllerAtIndex(int index) async {
+    final controller = _controllers[index];
+    if (widget.shortsVideos.length > index && index >= 0 && controller != null) {
```

Comment:
Do you see how you can refactor these 3 functions?

Reply:
DRY

Reply:
I just try to made it into one validation function